akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSubscriber.scala

```diff
@@ -0,0 +1,119 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl.io
+
+import akka.actor.{ ActorLogging, ActorRef, Deploy, Props }
+import akka.stream.actor.{ ActorSubscriberMessage, RequestStrategy }
+import akka.util.ByteString
+
+import scala.concurrent.Promise
+
+/** INTERNAL API */
+object InputStreamSubscriber {
+  def props(completionPromise: Promise[Long], bufSize: Int): Props = {
+    require(bufSize > 0, "buffer size must be > 0")
+    Props(classOf[InputStreamSubscriber], completionPromise, bufSize).withDeploy(Deploy.local)
+  }
+
+  trait InputStreamSubscriberInMessage
+
+  case class Read(length: Int) extends InputStreamSubscriberInMessage
+
+  case class Close() extends InputStreamSubscriberInMessage
+
+  trait InputStreamSubscriberOutMessage
+
+  case class Bytes(bytes: ByteString) extends InputStreamSubscriberOutMessage
+
+  case class Failure(msg: String) extends InputStreamSubscriberOutMessage
+
+}
+
+/** INTERNAL API */
+private[akka] class InputStreamSubscriber(bytesWrittenPromise: Promise[Long], bufSize: Int)
+  extends akka.stream.actor.ActorSubscriber
+  with ActorLogging {
+
+  var bytesRequestedDownstream: Option[(ActorRef, Int)] = None
+
+  var bytesWritten: Long = 0
+
+  var availableChunks: Vector[ByteString] = Vector.empty[ByteString]
+
+  override protected val requestStrategy = new RequestStrategy() {
+    override def requestDemand(remainingRequested: Int): Int = bytesRequestedDownstream match {
+      case Some((actor, length)) ⇒ if (remainingRequested > 0) 0 else 1
```

Comment:
random trivia, until recently scala allocates a lot of tuples for these extractors (2.12 improves on that AFAIR).
Not that it hurts much here, since we're IO dominated anyway